### ISSUE TITLE
Streamlining for P4_16

### ISSUE THREAD

**Issue by cc10512** [p4lang/p4c#363]
moving from: https://github.com/p4lang/p4-spec/issues/125

@mbudiu-vmw I understand your concerns. It has worked well for a small number of developers, but we're finding that as we bring on new folks, it is becoming an issue since there is no clear history and separation for some of these features.

As I mentioned yesterday, we would like to get the compiler to the point that it supports all the features we're discussing in the WG in time for P4 developer day. This is a very aggressive schedule, and having to debug and support experimental features in the main branch is adding costs that we can't really afford. I'm open to suggestions on how we can reduce the burden.

**Comment by sethfowler** [comment]
For web standards, the usual approach is that whenever possible features are implemented before standardization as "polyfills" - effectively, library code that adds support for the feature - or in source-to-source translators. This reduces the burden on implementers and makes it possible to experiment extensively with new features before standardizing them and putting lots of effort into an full implementation.

Where possible, it would be preferable to take that approach with experimental features in p4c. There's already some precedent for this: we internally translate P4-14 code to P4-16. Ideally, once this translation happens, code later in the pipeline should not need to know anything about P4-14 constructs.

We could extend this approach to other experimental features - for example, generics. This would reduce the maintenance burden we incur to support them, and it would reduce the amount of complexity that people working on the compiler need to learn and manage.

We can probably do this while retaining much of the existing code that implements these features.

**Comment by jnfoster** [comment]
@sethfowler 

It's certainly possible to have experimental code side-by-side with standard code, but things are not very clean right now. Let me share two concrete examples that tripped me today: (i.e., I lost more than a few minutes one each, in parts of the compiler that I'm familiar with):

1. The IR allows Declaration_Instances to an initializer which can be null. This is an experimental feature, but it's not documented in the IR (https://github.com/p4lang/p4c/blob/master/ir/ir.def#L274). Instead, one has to look at the sub-production in the parser where there is a comment flagging this as experimental (https://github.com/p4lang/p4c/blob/master/frontends/p4/p4-parse.ypp#L329). 

2. The parser recognizes function *declarations* as expressions. This is not documented in the IR (https://github.com/p4lang/p4c/blob/master/ir/ir.def#L402) or the parser (https://github.com/p4lang/p4c/blob/master/frontends/p4/p4-parse.ypp#L857). I had to use Git blame to see when it was added, and deduce from the commit message that it's an experimental feature (https://github.com/p4lang/p4c/blame/master/frontends/p4/p4-parse.ypp#L857).

Long term, of course, we should have a way to develop experimental features (certainly on branches) and incorporate them into master (properly documented, delineated, and enabled with the appropriate flags). I agree that things that can be handled in the front-end should have minimal impact. And the browser "polyfill" approach sounds attractive. But, in the short-term, I am **seriously concerned** that we will be unable to get all of the bugs out of the front-end while maintaining experimental features, at least in their current form. Moreover, even if we were to adopt all of the changes that were proposed on Monday, I don't think we'd end up relegating very much code to branches -- e.g. we will still need an implementation of Hindley-Milner since externs have generics.

So my strong preference would be to agree to relegate most or all experimental features to branches in the short term and then build back up from a solid foundation. 

By getting the implementation in to a clean, lean state, we will gain momentum and create a good platform for experimenting with new ideas in the future.

**Comment by mbudiu-vmw** [comment]
1. I am not aware of evidence that the existing experimental features are slowing us down. There are no outstanding issues filed against these as far as I can tell.

2. All of the experimental features, with the exception of header_unions, were added at the request of Barefoot, to support extensions to P4 v1.0 and P4 v1.1 which were undocumented.

3. By far the most useful thing we can do to remove bugs is to write more tests. We have very few tests, we should have tens of thousands, and we have even fewer end-to-end tests using BMv2. We have received very few test contributions.

4. Adding relatively significant major language changes as proposed recently in the design group won't really help reduce the bug count, but it may create significant additional work.

5. I don't know how you plan to move the experimental features to branches. I propose that someone tries to move one of the features to a branch, to see how much work is involved. I personally am not good enough with git and automake to even contemplate doing this. Running tests takes a significant amount of time, and I have no idea how you would test many branches.

**Comment by sethfowler** [comment]
Good news: we actually already run tests against all the branches in this repo! You can check on them in the Travis UI.

Regarding point #1, the biggest concern here is the maintenance burden imposed by these features, and that won't necessarily show up in the form of Github issues directly complaining about the feature. It will usually show up in more subtle ways - bugs that are introduced in part because the code is hard to follow, time wasted trying to do code archaeology to understand why something is implemented in an unexpected way, or just plain extra work to handle cases that we wouldn't otherwise need to handle. Maintenance burden is something that is extremely real, but notoriously hard to quantify.

Tests are always a fantastic idea, and the more the better. To my mind, though, there's another task which is just as important: we need to document the code and what features it's supposed to be implementing. That way, we'll know what needs testing, and we will stop struggling to implement the bug fixes and new features because we don't understand our own codebase. It's likely that the process of producing that documentation will also give us a lot of ideas for new tests to write.

**Comment by mbudiu-vmw** [comment]
I agree with code documentation; there are two bits of documentation which I tried to maintain relatively up-to-date: the spec and the compiler-design.pptx file in docs/. The latter is a bit out of date but could be refreshed without too much effort; the part describing the build system is the most obsolete, since I am not really keeping track of all the changes you guys have done lately, starting with the unified build and the new testing framework. (There are also some comments in the code, but they are somewhat sparse.)
